CAIPs/caip-211.md

```diff
@@ -0,0 +1,275 @@
+mean the wallet either does not support custom endpoints/definitions or the user
+has opted out of that degree of trust, which for privacy/fingerprinting reasons,
+should not be handled by distinct codepaths (See Privacy Considerations).
+
+## Specification
+
+Two properties are added to the scope objects requested in `optionalScopes` and
+optionally returned in `sessionScopes`.  These are both [ordered] strings of arrays:
+
+1. `rpcEndpoints` is an array of zero or more URLs of RPC endpoints that
+   the caller would prefer the respondent to use, ordered by preference. Each
+   must be a valid URL that addresses an RPC endpoint. The respondent may
+   return it empty, reordered, with less, the same, or even more conformant URLs
+   than received.
+2. `rpcDocuments` is an array of zero or more URLs of machine-readable RPC
+   documents that the caller would prefer the respondent to use, ordered by
+   preference. This set of document collectively defines at least syntactically
+   if not also semantically any methods and/or notifications authorized by the
+   CAIP-25 authorization, in DESCENDING heirarchical authority. (For example,
+   any methods or notifications defined differently by multiple authorites will
+   be interpreted by whichever authority is closer to the 0-index of the array).
+   Each must be a valid URL that addresses a valid openRPC document. The
+   respondent may return it empty, reordered, with less, the same, or even more
+   conformant URLs than received.
+
+### Request
+
+A CAIP-211 request is a valid CAIP-25 except for the two additional properties. 
+
+Example:
+
+```jsonc
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "method": "provider_authorization",
+  "params": {
+    "requiredScopes": {
+      "eip155": {
+        "chains": ["eip155:1", "eip155:137"],
+        "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_sign", "get_balance", "personal_sign"],
+        "notifications": ["accountsChanged", "chainChanged"]
+      },
+      "eip155:42069": {
+        "methods": ["get_balance", "chainChanged", "42069_sEcReTbAlAnCe"],
+        "rpcDocuments": ["https://openrpc.42069-chain.org/", "https://raw.githubusercontent.com/ethereum/execution-apis/assembled-spec/refs-openrpc.json"],
+        "rpcEndpoints": ["https://node1.42069-chain.org/"]
+      },
+      "cosmos": {
+        ...
+      }
+    },
+    "optionalScopes":{
+      "eip155:42161": {
+        "methods": ["eth_sendTransaction", "eth_signTransaction", "get_balance", "personal_sign"],
+        "events": ["accountsChanged", "chainChanged"]
+    },
+    "sessionProperties": {
+      "expiry": "2022-12-24T17:07:31+00:00",
+      "caip154-mandatory": "true"
+    }         
+  }
+}
+```
+
+### Response
+
+The wallet can respond to this method with either a success result or an error message.
+
+#### Success
+
+A succesful result will be a conformant successful result according to
+[CAIP-25][], with the additional presence (if authorized) of the new properties
```

Comment:
I don't know if it makes sense to include the new properties in the response, what use it to spit back what the dapp has requested from you? The OpenRPC doc would be just used by the wallet for discovery.

Reply:
my thinking was that the ORDERING of the array might be changed by the wallet (i.e., I'll trust this RPC extension doc, but wherever it redefines something in my core doc, e.g. execution-apis, I'll take the latter's definition). This allows for some flexibility between binary accepting/rejecting each element in the array.

Reply:
On further thought, it might make sense to include the `rpcEndpoints` field in an ordered manner like you mentioned for `rpcDocuments`. If a provider so chooses to use the provided endpoints, it can specify usage + fallbacks in order or otherwise return an empty array (if its using endpoints of its own choice). 

Ideally,  there should only be one source of truth for the `rpcDocument`, lets call it that because well it should be one. Imagine a scenario where for some reason the multiple endpoints are not updated to be in line with each other, we have a scenario where the wallet doesn't know what document to trust. Anytime a request has an OpenRPC doc, the provider should be using that to determine/provide functionality, including for core-apis.

Reply:
Good point.  I guess I was assuming that multiple RPC docs don't necessarily undermine THE RPC Doc.  

Example: custom endpoints might find it useful to have RPCDocs reinforcing or making more explicit what's custom about them-- whether the custom RPC doc overrides The One is a matter of policy, easily expressed in ordering? Why not let additional RPC docs *extend* THE RPC Doc, like a class?

Maybe that's too footgunny, though.  The real question is whether that ordering mechanism is too prone to human error or too hackable... let's talk about it at the next meeting and see what others in the group have to say, maybe? Feel free to bring Shane or anyone at MM who has opinions and/or topical devrel experience from EthDenver :D

Reply:
See also this exchange with ligi, which refers to a discussion with TimDaub about eth clients at the last editorial meeting: 
https://github.com/ChainAgnostic/CAIPs/pull/206#discussion_r1102520837

It seems multiple *only very slightly different* RPC documents might be useful until there is 100% conformance across all eth clients!  I can definitely imagine that on other namespaces (e.g. Polkadot) where chains can customize their runtimes and add crates/pallets, it would be very useful for cross-chain apps to have multiple RPC docs...

Reply:
Perhaps namespace-specific profiles of 211 would be the place to recommend (or even constrain) behavior here, like "never authorize more than 1 RPC Doc" or "Never put any RPC ahead of The One when authorizing"?